test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -60,6 +85,27 @@
         ;; make sure we got here
         (is (true? true))))))
 
+(deftest pool-lock-is-blocking-test-until-borrows-unregistered-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 2)
+        future-started? (promise)]
+    (is (= 2 (count instances)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool))]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are unregistered"
+        (is (not (realized? lock-thread)))
+
+        (testing (str "other threads may successfully unregister instances "
+                      "while pool.lock() is being executed")
+          (.unregister pool (first instances))
+          (is (not (realized? lock-thread)))
+          (.unregister pool (second instances)))
+
+        @lock-thread
+        ;; make sure we got here
+        (is (true? true))))))
```

Comment:
Or maybe a better current case would be around the full JRuby pool flush.  In that case, if the lock were requested against the soon-to-be "old" pool, the lock() would be held off until all instances were back in the "old" pool.  I think the logic in the pool flush will borrow the instance from the "old" pool, unregister it without putting it back into the "old" pool, create a replacement instance, and put the replacement instance in the "new" pool.  There is some logic in the pool flush to put a poison pill in the "old" pool, but it only does one rather than one per instance being disposed.  Maybe it would have been better to do one pill per instance being disposed?  In any event, I think the logic to awaken the lock requester after all of the pool instances have been unregistered would prevent the lock request from otherwise getting stuck indefinitely.